Added --enable-metrics support in srt-xtransmit (#39)
mbakholdina/lib-srt-utils: configs/README.md

# Experiment Config Demystified

## Experiment Config and Task Concept

> Experiment config is a `.json` file consisting of the following fields:

Reviewer: @mbakholdina Is "Experiment config" the name of the file?

Author: Here it's in general any config file.

> Experiment config is a `.json` file consisting of the following fields:

Reviewer: ```suggestion
An experiment config is a `.json` file consisting of the following fields:
```

> - `collect_results_path`: Directory path to collect experiment results.

Reviewer: ```suggestion
- `collect_results_path`: path to the directory where experimental results are collected.
```

- `stop_after`: Time after which an experiment will be stopped, in seconds.
- `ignore_stop_order`: `true` if stopping order of tasks will be ignored when stopping an experiment.
- `tasks`: Tasks to run within an experiment and their configs listed in starting order.

```
{
    "collect_results_path": "_experiment_results",
    "stop_after": 70,
    "ignore_stop_order": true,
    "tasks": {
        "1": {...},
        "2": {...},
        ...
    }
}
```

> A task represents one step of a single experiment and contains both information about an object to run (e.g., `thark` or `srt-xtransmit` application) and a way to run the object (e.g., locally or on a remote machine). Task config also contains information about the time to sleep after the object start/stop and stopping order. However, those fields are optional and should be set to `null` if not defined.

Reviewer: ```suggestion
A task represents one step of a single experiment and contains both information about an object to run (e.g., `thark` or `srt-xtransmit` application) and where to run the object (e.g., locally or on a remote machine). A task config also contains information about the time to sleep after the object starts or stops, and the stopping order. However, those fields are optional and should be set to `null` if not defined.
```

Here is an example of task config:

```
"1": {
    "obj_type": "srt-xtransmit",
    "obj_config": {...},
    "runner_type": "remote-runner",
    "runner_config": {...},
    "sleep_after_start": null,
    "sleep_after_stop": null,
    "stop_order": null
}
```

where

- `obj_type`: Object to run. Currently `tshark` and `srt-xtransmit` applications are supported.
- `obj_config`: Object config that helps to build a command line for a Python subprocess.
- `runner_type`: Object runner to use. `local-runner` should be used to run the object locally and `remote-runner` - on a remote machine via SSH.
- `runner_config`: Object runner config.
- `sleep_after_start`: Time to sleep after object start, in seconds. Optional, must be `null` if not defined.
- `sleep_after_stop`: Time to sleep after object stop, in seconds. Optional, must be `null` if not defined.

> - `stop_order`: The stopping order of tasks. Optional, must be `null` if not defined. By default, tasks will be stopped in reversed order.

Reviewer: ```suggestion
- `stop_order`: The stopping order of tasks. Optional, must be `null` if not defined. By default, tasks will be stopped in reverse order.
```

## Experiment Description

> The usual experiment consists of the following tasks:
> 1. Start `tshark` locally or remotely via SSH on a receiver side. Optional.
> 2. Start `tshark` locally or remotely via SSH on a sender side. Optional.
> 3. Start `srt-xtransmit` test application (or another test application with SRT support), receiver, locally or remotely via SSH.
> 4. Start `srt-xtransmit` test application (or another test application with SRT support), sender, locally or remotely via SSH.

Reviewer: ```suggestion
A typical experiment consists of the following tasks:
1. Start `tshark` locally or remotely via SSH on a receiver (optional).
2. Start `tshark` locally or remotely via SSH on a sender (optional).
3. Start `srt-xtransmit` test application (or another test application with SRT support), receiver, locally or remotely via SSH.
4. Start `srt-xtransmit` test application (or another test application with SRT support), sender, locally or remotely via SSH.
```

> Adding or deleting tasks under the `tasks` field of the experiment config changes logic of the experiment.

Reviewer: ```suggestion
Adding or deleting tasks under the `tasks` field of an experiment config changes the logic of the experiment.
```

> Once all the tasks have been started, script sleeps for `stop_after` seconds to wait for the SRT sender to finish transmission and then stops the experiment.

Reviewer: ```suggestion
Once all the tasks have been started, the script sleeps for `stop_after` seconds to wait for the SRT sender to finish transmission and then stops the experiment.
```

> After stopping the tasks, script collects experiment artifacts (SRT `.csv` statistics, a file with metrics produced by the `srt-xtransmit` receiver, `.pcapng` dumps generated by `tshark`) to machine where the script is running.

Reviewer: ```suggestion
After stopping the tasks, the script collects experiment artifacts (SRT `.csv` statistics, a file with metrics produced by the `srt-xtransmit` receiver, `.pcapng` dumps generated by `tshark`) and writes them to the machine where the script is running.
```

"runner_type": "remote-runner",
"runner_config": {
    "username": "msharabayko",
    "host": "10.129.10.92"
},
```

> - `username`: Username on the remote machine to connect through.

Reviewer: ```suggestion
- `username`: Username on the remote machine through which the script connects.
```

> - `host`: IP address of the remote machine to connect.

Reviewer: ```suggestion
- `host`: IP address of the remote machine involved in the experiment.
```
